### ISSUE TITLE
Pipeline stuck in "pending" state

### ISSUE THREAD

**Issue by ringods** [concourse/concourse#6615]
## Summary

We have a pipeline which has a full pipeline in `pending` state.

## Steps to reproduce

Hardly possible to deterministically reproduce unfortunately.

## Expected results

Basic `check` functionality should work in all cases.

## Actual results

Pipeline in `pending` state.

## Additional context

Stack dump of `web` and 2 `worker` machines:
https://gist.github.com/ringods/35d9011d7a1b8199f2a4abbb500ac978

## Triaging info

* Concourse version: 7.0.0
* Did this used to work?
Not always. The aggregate issue on this is #844. Followed [the advice](https://github.com/concourse/concourse/issues/844#issuecomment-496623549) to get a stack dump and create a new issue.

**Comment by taylorsilva** [comment]
Confused by this:
> We have a pipeline which has a full pipeline in pending state.

Do you mean you have a pipeline where every **job** is stuck in pending?

**Comment by ringods** [comment]
@taylorsilva indeed. Sorry that I did not re-read my original post carefully. 

Meanwhile we found out that some of the `check` containers were probably still "linked" to stalled workers, even if these workers were already for days in that state. Is there a specific reason why Concourse doesn't reschedule to an active worker?

After cleaning up the stalled workers, the pipelines resumed working. We also reconfigured our workers with the `--ephemeral` flag to prevent this situation of happening again.

**Comment by taylorsilva** [comment]
> Is there a specific reason why Concourse doesn't reschedule to an active worker?

No particular reason, this could probably be improved upon. The general idea with stalled workers is that they "may" come back and therefore we would want to continue using their volumes and containers. Like maybe the connection dropped for 30s and then recovered. No point in deleting all those workloads, just keep using them!

It's probably reasonable to add a timeout for how long Concourse waits for a stalled worker to return before considering it gone, along with all its volume and containers. Currently Concourse will wait indefinitely until you `fly prune-worker`. Definitely some rough edges here.

**Comment by clarafu** [comment]
I think before we can do any investigation into this issue, we will need more information as to what exactly the pipeline was stuck at during it's `pending` state. When you saw the build within the pipeline stuck in pending state, did it say where it was stuck at in the preparation page (everything was checked)? If you check the web logs if you grep for that job name that had the pending build, did it have any suspicious logs? When you say there were `check` containers still linked to stall workers, do you mean that there are check containers still running a check on those stalled workers or just that when you run `fly containers` you saw there were still check containers linked to those stalled workers.

**Comment by piroux** [comment]
> Meanwhile we found out that some of the check containers were probably still "linked" to stalled workers, even if these workers were already for days in that state. Is there a specific reason why Concourse doesn't reschedule to an active worker?

This rings a :bell: , as I think I have experienced that as well !

I was running Concourse `v6.5.1` with docker-compose and a single worker (without the `--ephemeral` flag), for around 100 pipelines.
I had to restart Concourse and reboot the server on a daily basis because of volume leakage issues which were filling up the disk. Initially I was not pruning the workers from previous restarts, so that Concourse was still keeping track of those "stalled" workers and maybe assigning (check) containers to one of the missing and stalled workers, I don't really remember exactly. And usually after a week of doing so, most jobs were failing to start by not finding the single running worker. So I solved it by pruning the workers on a weekly basis, then daily basis. In the end, with this workflow, it was working well enough.
Unfortunately I have not kept the logs.

Recently, I upgraded to Concourse `v6.7.6` and I found out that there were no more volume leakage issues filling up the disk so that I stopped restarting Concourse periodically altogether and everything is working well.
If this issue returns later on, I will come back here with more details.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
